<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">

    <head>
        <title>ChIPmentation: fast, cheap, low-input ChIP-seq for histones and transcription factors</title>
        <meta name="ROBOTS" CONTENT="NOINDEX, NOFOLLOW"/>
        <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
        <link href="style.css" rel="stylesheet" type="text/css">
    </head>

    <body>

        <h2>ChIPmentation: fast, cheap, low-input ChIP-seq for histones and transcription factors</h2>

        <h3>Abstract</h3>

        <p>
        Chromatin immunoprecipitation followed by next generation sequencing (ChIP-seq) is widely used to map histone marks and transcription factor binding throughout the genome.Here we present ChIPmentation, a method that combines chromatin immunoprecipitation with sequencing library preparation by Tn5 transposase enzyme (“tagmentation”). ChIPmentation introduces sequencing-compatible adapters in a single-step reaction directly on bead-bound chromatin, which reduces time, cost, and input requirements, thus providing a convenient and broadly useful alternative to existing ChIP-seq protocols.
        </p>

        <h3>Data</h3>
        <p>
        52 libraries were sequenced for ChIPmentation, 24 libraries for standard ChIP-seq, nine libraries for ChIP-tagmentation, and two libraries for ATAC-seq.
        </p>
        <p>
            <a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&hubClear=http://biomedical-sequencing.at/bocklab/papers/schmidl2015/hub.txt&position=chr16:88485000-88885000">UCSC Genome Browser track hub</a>
        </p>
        <p>
            <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70482">GEO accession number GSE70482</a>
        </p>

        <!--
        <h3>News</h3>
        <p></p>
        -->

        <h3>Protocols</h3>
        <table width="1100" border="1">
            <thead>
                <tr>
                    <td valign="top"><strong></strong></td>
                    <td valign="top"><strong>Protocol version</strong></td>
                    <td valign="top"><strong>Updated on</strong></td>
                    <td valign="top"><strong>Tested with</strong></td>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td valign="top"><strong>ChIPmentation</strong></td>
                    <td valign="top"><a href="protocols/ChIPmentation_protocol.v1.0.pdf"> v1.0 (publication)</a></td>
                    <td valign="top">7th July 2015</td>
                    <td valign="top">
                        H3K4me1 (10M cells), H3K4me3 (10-500k cells), H3K27ac (10M cells), H3K27me3 (10k-500k cells), H3K36me3 (10M cells),
                        <br>
                        CTCF (100k-10M cells), GATA1 (100k-10M cells), PU.1 (500k-10M cells), REST (100k-10M cells)
                    </td>
                </tr>
            </tbody>
        </table>

        <!--
        CTCF (Millipore 07-729)
        GATA1 (Abcam ab11852)
        H3K27ac (Diagenode pAb-196-050)
        H3K27me3 (Millipore 07-449)
        H3K36me3 (Diagenode pAb-192-050)
        H3K4me1 (Diagenode pAb- 194-050)
        H3K4me3 (Diagenode pAb-003-050)
        PU.1 (Santa Cruz sc-352)
        REST (Millipore 07-579)
        -->

        <!--
        <h3>Publications using ChIPmentation</h3>
        <table width="1100" border="1">
            <thead>
                <tr>
                    <td valign="top"><strong></strong></td>
                    <td valign="top"><strong>Title</strong></td>
                    <td valign="top"><strong>Cited on</strong></td>
                    <td valign="top"><strong>doi</strong></td>
                </tr>
            </thead>
            <tbody>
                <tr>
                </tr>
            </tbody>
        </table>
        -->

        <h3>Citation</h3>
        <p>
            If you use the ChIPmentation protocol, please cite:
        </p>
        <p>
            Christian Schmidl, André F. Rendeiro, Nathan C. Sheffield & Christoph Bock.<br />
            <i>ChIPmentation: fast, cheap, low-input ChIP-seq for histones and transcription factors</i><br />
            Nature Methods, 2015 doi:<a href="http://dx.doi.org/"> </a>
        </p >


        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

            ga('create', 'UA-65022614-1', 'auto');
            ga('send', 'pageview');
        </script>

    </body>

</html>
